`[pallet_xcm::reserve_transfer_assets]` `NotHoldingFees` issue in case of paid XcmRouter is used

bkontur opened the pull request:
Based on PR: https://github.com/paritytech/polkadot/pull/7005

on AssetHubs we plan to set FeeManager and waive unpaid only for system parachains or relay chain
https://github.com/paritytech/cumulus/pull/2433/files#diff-9a51956f7781a349d207a9d5f54ea0061666a856fb108d1a1190bd5178eaac50R432-R437
so it means, that any user/origin will be charged for delivery, from [holding register](https://github.com/paritytech/polkadot/blob/6f4a8bb0063c91fe869b94196ead5717950b9f23/xcm/xcm-executor/src/lib.rs#L399-L412):
```
                let (ticket, fee) = validate_send::<Config::XcmSender>(dest, msg)?;
		if !Config::FeeManager::is_waived(self.origin_ref(), reason) {
			let paid = self.holding.try_take(fee.into()).map_err(|_| XcmError::NotHoldingFees)?;
			Config::FeeManager::handle_fee(paid.into(), Some(&self.context));
		}
		Config::XcmSender::deliver(ticket).map_err(Into::into)
```
which seems ok, but there are several cases, when this `holding` register is not filled and is empty,
for example `pallet_xcm::reserve_transfer_assets` generates instructions:
```
SetFeesMode { jit_withdraw: true },
TransferReserveAsset { assets, ...}
```

`SetFeesMode` -> does nothing, because it affects only `sef.take_fee(` which is used only for `ExportMessage / LockAsset / RequestUnlock` 
`TransferReserveAsset` - does reserve transfer of assets and then `self.send(ReserveAssetDeposited(assets))` to destination

**and here are problems with this `self.send`:**

---
1. there is nothing in `holding` so this `self.send` for regular user fails on `NotHoldingFees` - see test bellow without fix
---
2. Should we withdraw `fee from validate_send`  from user account directly or from holding register?
maybe here in `fn send` is missing `IF` based on `jit_withdraw: true`, which is the same code as `fn take_fee` does?

so maybe solution is to use here `self.take_fee(` - which is working for test :)
```
/// Send an XCM, charging fees from Holding as needed.
	fn send(
		&mut self,
		dest: MultiLocation,
		msg: Xcm<()>,
		reason: FeeReason,
	) -> Result<XcmHash, XcmError> {
		let (ticket, fee) = validate_send::<Config::XcmSender>(dest, msg)?;
		if !Config::FeeManager::is_waived(self.origin_ref(), reason) {
			self.take_fee(fee, reason)?;
		}
		Config::XcmSender::deliver(ticket).map_err(Into::into)
	}
```





---
**Test fails withtout changing:**

orginal:
```
		if !Config::FeeManager::is_waived(self.origin_ref(), reason) {
			let paid = self.holding.try_take(fee.into()).map_err(|_| XcmError::NotHoldingFees)?;
			Config::FeeManager::handle_fee(paid.into(), Some(&self.context));
		}
```

fixed to:
```
		if !Config::FeeManager::is_waived(self.origin_ref(), reason) {
			self.take_fee(fee, reason)?;
		}
```





```
cargo test --package pallet-xcm --lib tests::reserve_transfer_assets_with_paid_router_works -- --exact

Left:  RuntimeEvent::XcmPallet(Event::Attempted { outcome: Incomplete(Weight { ref_time: 2000, proof_size: 2000 }, NotHoldingFees) })
Right: RuntimeEvent::XcmPallet(Event::Attempted { outcome: Complete(Weight { ref_time: 2000, proof_size: 2000 }) })
<Click to see difference>

thread 'tests::reserve_transfer_assets_with_paid_router_works' panicked at 'assertion failed: `(left == right)`
  left: `RuntimeEvent::XcmPallet(Event::Attempted { outcome: Incomplete(Weight { ref_time: 2000, proof_size: 2000 }, NotHoldingFees) })`,
 right: `RuntimeEvent::XcmPallet(Event::Attempted { outcome: Complete(Weight { ref_time: 2000, proof_size: 2000 }) })`', xcm/pallet-xcm/src/tests.rs:563:9
```

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/3340204

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/3341419

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/3346534

KiChjang reviewed (approved):
I _think_ the intention is that _any_ fees payable in the XCM executor can be configured via `SetFeesMode`, but some of the existing code were not changed to respect the `fees_mode` register, so this change LGTM.

gavofyork commented on runtime/kusama/src/lib.rs:
```diff
@@ -2130,6 +2132,43 @@ sp_api::impl_runtime_apis! {
 						fun: Fungible(1_000_000 * UNITS),
 					}].into()
 				}
+				fn ensure_for_send(origin_ref: &MultiLocation, _dest: &MultiLocation, fee_reason: FeeReason) -> Option<FeesMode> {
+					use xcm_executor::traits::{FeeManager, TransactAsset};
+
+					let mut fees_mode = None;
+					if !<XcmConfig as xcm_executor::Config>::FeeManager::is_waived(Some(origin_ref), fee_reason) {
+						// if not waived, we need to set up accounts for paying and receiving fees
+
+						// mint ED to origin
+						let ed = MultiAsset {
+							id: Concrete(TokenLocation::get()),
+							fun: Fungible(EXISTENTIAL_DEPOSIT.into())
+						};
+						<XcmConfig as xcm_executor::Config>::AssetTransactor::deposit_asset(&ed, &origin_ref, None).unwrap();
+
+						// overestimate delivery fee
+						use runtime_common::xcm_sender::PriceForParachainDelivery;
+						let overestimated_xcm = vec![ClearOrigin; 128].into();
+						let overestimated_fees = runtime_common::xcm_sender::ExponentialPrice::<
+							xcm_config::FeeAssetId,
+							xcm_config::BaseDeliveryFee,
+							TransactionByteFee,
+							Dmp
+						>::price_for_parachain_delivery(
+							kusama_runtime_constants::system_parachain::ASSET_HUB_ID.into(),
+							&overestimated_xcm
+						);
+
+						// mint overestimated fee to origin
+						for fee in overestimated_fees.inner() {
+							<XcmConfig as xcm_executor::Config>::AssetTransactor::deposit_asset(&fee, &origin_ref, None).unwrap();
+						}
+
+						// expected worst case
+						fees_mode = Some(FeesMode { jit_withdraw: true });
+					}
+					fees_mode
+				}
```
This (rather complex) code will likely need to be copy-pasted across every single Polkadot/Kusama chain. Therefore, best to keep it in some common place and have a simple means of calling into it with some appropriate parameterisation.

bkontur replied:
ok, thank you, extracted to reusable code, which will be also reused for system parachains

gavofyork commented on runtime/polkadot/src/lib.rs:
```diff
@@ -2108,6 +2110,43 @@ sp_api::impl_runtime_apis! {
 					// Polkadot only knows about DOT
 					vec![MultiAsset { id: Concrete(TokenLocation::get()), fun: Fungible(1_000_000 * UNITS) }].into()
 				}
+				fn ensure_for_send(origin_ref: &MultiLocation, _dest: &MultiLocation, fee_reason: FeeReason) -> Option<FeesMode> {
+					use xcm_executor::traits::{FeeManager, TransactAsset};
+
+					let mut fees_mode = None;
+					if !<XcmConfig as xcm_executor::Config>::FeeManager::is_waived(Some(origin_ref), fee_reason) {
+						// if not waived, we need to set up accounts for paying and receiving fees
+
+						// mint ED to origin
+						let ed = MultiAsset {
+							id: Concrete(TokenLocation::get()),
+							fun: Fungible(EXISTENTIAL_DEPOSIT.into())
+						};
+						<XcmConfig as xcm_executor::Config>::AssetTransactor::deposit_asset(&ed, &origin_ref, None).unwrap();
+
+						// overestimate delivery fee
+						use runtime_common::xcm_sender::PriceForParachainDelivery;
+						let overestimated_xcm = vec![ClearOrigin; 128].into();
+						let overestimated_fees = runtime_common::xcm_sender::ExponentialPrice::<
+							xcm_config::FeeAssetId,
+							xcm_config::BaseDeliveryFee,
+							TransactionByteFee,
+							Dmp
+						>::price_for_parachain_delivery(
+							polkadot_runtime_constants::system_parachain::ASSET_HUB_ID.into(),
+							&overestimated_xcm
+						);
+
+						// mint overestimated fee to origin
+						for fee in overestimated_fees.inner() {
+							<XcmConfig as xcm_executor::Config>::AssetTransactor::deposit_asset(&fee, &origin_ref, None).unwrap();
+						}
+
+						// expected worst case
+						fees_mode = Some(FeesMode { jit_withdraw: true });
+					}
+					fees_mode
+				}
```
Oh look... there it is again already.

gavofyork reviewed (changes requested):
Overall solution seems sound, however too much code duplication and bare logic is introduced into high-level runtime definition code which should be 99% declarative.

Something for all Frame coders to note: 

The central `lib.rs` files of runtimes should not contain bare logic. The point of the extensive use of traits and types in how I designed Frame is to avoid exactly this. Runtimes should *ONLY CONTAIN BASIC TYPE DEFINITIONS*. Any logic should be parcelled up into one or more reusable types and stashed somewhere common for everyone else to make use of too.

bkontur commented:
bot rebase

bkontur commented:
bot rebase

bkontur commented:
bot rebase

bkontur commented:
bot fmt

bkontur commented:
> Overall solution seems sound, however too much code duplication and bare logic is introduced into high-level runtime definition code which should be 99% declarative.
> 
> Something for all Frame coders to note:
> 
> The central `lib.rs` files of runtimes should not contain bare logic. The point of the extensive use of traits and types in how I designed Frame is to avoid exactly this. Runtimes should _ONLY CONTAIN BASIC TYPE DEFINITIONS_. Any logic should be parcelled up into one or more reusable types and stashed somewhere common for everyone else to make use of too.

@gavofyork I fixed and refactored those duplication like you suggested, could you please, check? I hesitate to merge without your explicit approve here (because you requested changes), thank you :)

franciscoaguirre commented on xcm/pallet-xcm/src/mock.rs:
```diff
@@ -193,6 +195,38 @@ impl SendXcm for TestSendXcmErrX8 {
 	}
 }
 
+parameter_types! {
+	pub Para3000: u32 = 3000;
+	pub Para3000Location: MultiLocation = Parachain(Para3000::get()).into();
+	pub Para3000PaymentAmount: u128 = 1;
+	pub Para3000PaymentMultiAssets: MultiAssets = MultiAssets::from(MultiAsset::from((Here, Para3000PaymentAmount::get())));
+}
+/// Sender only sends to `Parachain(3000)` destination requiring payment.
+pub struct TestPaidForPara3000SendXcm;
```
It feels like these items are specific to a particular test. Why not put them in `tests` rather than in mock? The test sender should be generic enough.

franciscoaguirre reviewed (approved):
Just some nits

gavofyork commented on runtime/kusama/src/xcm_config.rs:
```diff
@@ -346,7 +348,8 @@ impl xcm_executor::Config for XcmConfig {
 	type MaxAssetsIntoHolding = MaxAssetsIntoHolding;
-	type FeeManager = XcmFeesToAccount<Self, SystemParachains, AccountId, TreasuryAccount>;
+	type FeeManager =
+		XcmFeesToAccount<Self, (SystemParachains, HereLocation), AccountId, TreasuryAccount>;
```
Please explain.

gavofyork reviewed (commented):
One query.
